src/cli/repl.rs

```diff
@@ -39,64 +43,218 @@ impl Validator for InputValidator {
-/// that's equal or greater than the number of iterations
+/// `pad(a, m)` returns the first multiple of `m` that's equal or greater than `a`
 ///
-/// Panics if reduction count is zero
-#[allow(dead_code)]
-fn pad_iterations(iterations: usize, rc: usize) -> usize {
-    let lower = rc * (iterations / rc);
-    if lower < iterations {
-        lower + rc
+/// Panics if `m` is zero
+#[cfg(not(target_arch = "wasm32"))]
+fn pad(a: usize, m: usize) -> usize {
+    let lower = m * (a / m);
+    if lower < a {
+        lower + m
     } else {
         lower
     }
 }
 
-impl<F: LurkField + serde::Serialize + for<'de> serde::Deserialize<'de>, C: Coprocessor<F>>
-    Repl<F, C>
-{
-    pub fn new(store: Store<F>, env: Ptr<F>, limit: usize, rc: usize) -> Result<Repl<F, C>> {
-        check_non_zero("limit", limit)?;
-        check_non_zero("rc", rc)?;
+#[cfg(not(target_arch = "wasm32"))]
+fn timestamp() -> u128 {
+    use std::time::{SystemTime, UNIX_EPOCH};
+    SystemTime::now()
+        .duration_since(UNIX_EPOCH)
+        .expect("We're after UNIX_EPOCH")
+        .as_nanos()
+}
+
+type F = pasta_curves::pallas::Scalar; // TODO: generalize this
+
+impl Repl<F> {
+    pub fn new(
+        store: Store<F>,
+        env: Ptr<F>,
+        limit: usize,
+        rc: usize,
+        backend: Backend,
+    ) -> Result<Repl<F>> {
+        validate_non_zero("limit", limit)?;
+        validate_non_zero("rc", rc)?;
+        let field = F::FIELD;
+        backend.validate_field(&field)?;
+        info!("Launching REPL with backend {backend} and field {field}");
         Ok(Repl {
             store,
             env,
             limit,
-            lang: Arc::new(Lang::<F, C>::new()),
-            // last_claim: None,
+            lang: Arc::new(Lang::new()),
             rc,
+            backend,
+            evaluation: None,
         })
     }
 
-    pub fn prove_last_claim(&mut self) -> Result<()> {
-        Ok(())
-        // match &self.last_claim {
-        //     Some(claim) => {
-        //         // TODO
-        //         let _proof = prove_claim(claim);
-        //         Ok(())
-        //     }
-        //     None => {
-        //         bail!("No claim to prove");
-        //     }
-        // }
+    #[cfg(not(target_arch = "wasm32"))]
+    pub fn prove_last_frames(&mut self) -> Result<()> {
+        use super::{
+            lurk_proof::LurkProofMeta,
+            paths::{proof_meta_path, proof_path},
+        };
+        use lurk::{
+            proof::{nova::NovaProver, Prover},
+            public_parameters::public_params,
+            z_store::ZStore,
+        };
+        use std::{fs::File, io::BufWriter};
+        match &self.evaluation {
+            None => bail!("No computation to prove"),
+            Some(Evaluation { frames, cost }) => match self.backend {
+                Backend::Nova => {
+                    // padding the frames
+                    let mut frames = frames.clone(); // don't mutate memoized frames
```

Comment:
We might want to add an option to interactively explore previous computations, like a live debug that allows the user to move back and forth.

Reply:
Right. How would the padding prevent that in any way?

Reply:
We can add an `iterations: usize` to prevent the interactive explorer from trespassing the limit. We don't want to visit padded frames there.

I tried adding it, but in the end I couldn't make it work by just mutating the frames in-place

Reply:
`NovaProver::prove` demands ownership of the frames, which belong to the `Repl`

Reply:
`NovaProver` didn't *really* need ownership of the frames, so I went there and changed it to receive a vector slice instead. The clone is now removed 🎉